File: compiler/bench/heappush.c
// expected: 352
#include <stdio.h>
#define N 100000

long heap[N];
int hsize = 0;

void swap_heap(int i, int j) {
    long t = heap[i];
    heap[i] = heap[j];
    heap[j] = t;
}

void sift_up(int i) {
    while (i > 0) {
        int p = (i - 1) / 2;
        if (heap[i] > heap[p]) {
            swap_heap(i, p);
            i = p;
        } else {
            break;
        }
    }
}

void push(long val) {
    heap[hsize] = val;
    sift_up(hsize);
    hsize++;
}

void init() {
    hsize = 0;
    for (int i = 0; i < N; i++) {
        push(((long)(i % 17) * (i % 23)) % N);
    }
}

int main() {
    init();
    printf("%ld\n", heap[0]);
    return 0;
}

File: compiler/bench/heappush.fs
\ expected: 352
\ Push 100K elements to heap, return max

100000 constant N
create heap N cells allot
variable hsize

: swap-heap ( i j -- )
  cells heap + swap cells heap +
  over @ over @
  rot ! swap ! ;

: sift-up ( i -- )
  begin
    dup 0> while
    dup 1- 2/
    over cells heap + @
    over cells heap + @
    > if
      2dup swap-heap
      nip
    else
      2drop exit
    then
  repeat
  drop ;

: push ( val -- )
  hsize @ cells heap + !
  hsize @ sift-up
  1 hsize +! ;

: init ( -- )
  0 hsize !
  N 0 do
    i 17 mod i 23 mod * N mod push
  loop ;

: main init heap @ . cr ;
